File: 04-computation/eulerian_n8_fourier_stress_overnight_hexagon_sep05.cpp
// Complete labelled cycle-space and dual-space enumeration, n<=8.
// No graph-isomorphism library, canonical-label heuristic, or sampling.
#include <algorithm>
#include <array>
#include <cstdint>
#include <iostream>
#include <numeric>
#include <stdexcept>
#include <vector>
using U = uint32_t;
using V = std::vector<int>;
using M = std::vector<V>;
void need(bool p, const char* s) { if (!p) throw std::runtime_error(s); }
struct Linear {
  std::array<std::array<U,128>,3> tab{};
  explicit Linear(const std::vector<U>& cols) {
    for(int b=0;b<3;++b) for(int x=1;x<128;++x) {
      int j=__builtin_ctz(unsigned(x)), k=7*b+j;
      tab[b][x]=tab[b][x^(1<<j)] ^ (k<int(cols.size())?cols[k]:0);
    }
  }
  U operator()(U x) const { return tab[0][x&127]^tab[1][(x>>7)&127]^tab[2][x>>14]; }
};
struct Orbits { V label; std::vector<std::vector<U>> members; };
Orbits quotient(U count,const std::vector<Linear>& maps) {
  Orbits o; o.label.assign(count,-1);
  for(U seed=0;seed<count;++seed) if(o.label[seed]<0) {
    int id=int(o.members.size()); o.members.push_back({seed}); o.label[seed]=id;
    auto& q=o.members.back();
    for(size_t k=0;k<q.size();++k) for(const auto& f:maps) {
      U y=f(q[k]); need(y<count,"linear range");
      if(o.label[y]<0) { o.label[y]=id; q.push_back(y); }
      else need(o.label[y]==id,"complete generator orbit");
    }
  }
  return o;
}
template<class T> void array_json(const std::vector<T>& v) {
  std::cout<<'['; for(size_t i=0;i<v.size();++i) { if(i) std::cout<<','; std::cout<<v[i]; } std::cout<<']';
}
void matrix_json(const M& a) {
  std::cout<<'['; for(size_t i=0;i<a.size();++i) { if(i) std::cout<<','; array_json(a[i]); } std::cout<<']';
}
int main(int argc,char** argv) {
  need(argc==2,"one argument n required"); int n=std::stoi(argv[1]); need(3<=n&&n<=8,"n in3..8");
  int idx[8][8], gi[8][8];
  for(auto& row:idx) std::fill(row,row+8,-1);
  for(auto& row:gi) std::fill(row,row+8,-1);
  std::vector<std::pair<int,int>> edges,gauge;
  for(int i=0;i<n;++i) for(int j=i+1;j<n;++j) { idx[i][j]=idx[j][i]=edges.size(); edges.push_back({i,j}); }
  for(int i=1;i<n;++i) for(int j=i+1;j<n;++j) { gi[i][j]=gi[j][i]=gauge.size(); gauge.push_back({i,j}); }
  int m=gauge.size(); U count=U(1)<<m;
  std::vector<U> basis;
  for(auto [i,j]:gauge) basis.push_back((U(1)<<idx[i][j])^(U(1)<<idx[0][i])^(U(1)<<idx[0][j]));
  std::vector<U> full(count,0);
  for(U x=1;x<count;++x) { int k=__builtin_ctz(x); full[x]=full[x^(U(1)<<k)]^basis[k]; }
  auto compress=[&](U x) { U a=0; for(int j=0;j<m;++j) if((x>>idx[gauge[j].first][gauge[j].second])&1) a|=U(1)<<j; return a; };
  std::vector<Linear> primal_maps,dual_maps;
  for(int v=0;v<n-1;++v) {
    V p(n); std::iota(p.begin(),p.end(),0); std::swap(p[v],p[v+1]);
    auto permute=[&](U x) { U a=0; for(int k=0;k<int(edges.size());++k) if((x>>k)&1) a^=U(1)<<idx[p[edges[k].first]][p[edges[k].second]]; return a; };
    std::vector<U> pc,dc;
    for(U x:basis) pc.push_back(compress(permute(x)));
    for(auto [a,b]:gauge) {
      U x=U(1)<<idx[p[a]][p[b]], h=0;
      for(int k=0;k<m;++k) {
        auto [i,j]=gauge[k];
        if(((x>>idx[i][j])^(x>>idx[0][i])^(x>>idx[0][j]))&1) h|=U(1)<<k;
      }
      dc.push_back(h);
    }
    primal_maps.emplace_back(pc); dual_maps.emplace_back(dc);
  }
  Orbits primal=quotient(count,primal_maps),dual=quotient(count,dual_maps);
  int q=primal.members.size(); need(q==int(dual.members.size()),"primal/dual class count");
  V order(q),relabel(q),sizes(q); std::vector<U> mins(q),reps(q),compressed(q);
  std::iota(order.begin(),order.end(),0);
  for(int i=0;i<q;++i) { mins[i]=UINT32_MAX; for(U x:primal.members[i]) mins[i]=std::min(mins[i],full[x]); }
  std::sort(order.begin(),order.end(),[&](int a,int b){return mins[a]<mins[b];});
  for(int i=0;i<q;++i) { relabel[order[i]]=i; reps[i]=mins[order[i]]; compressed[i]=compress(reps[i]); sizes[i]=primal.members[order[i]].size(); }
  for(int& i:primal.label) i=relabel[i];
  // An independent full-S_n path checks every canonical representative and
  // orbit size, including n8 (not just the generator implementation).
  V perm(n),stabilizers(q,0); std::iota(perm.begin(),perm.end(),0);
  int permutations=0;
  do {
    ++permutations;
    std::array<std::array<U,128>,4> tab{};
    for(int b=0;b<4;++b) for(int x=1;x<128;++x) {
      int j=__builtin_ctz(unsigned(x)),k=7*b+j;
      U bit=0;
      if(k<int(edges.size())) bit=U(1)<<idx[perm[edges[k].first]][perm[edges[k].second]];
      tab[b][x]=tab[b][x^(1<<j)]^bit;
    }
    for(int i=0;i<q;++i) {
      U r=reps[i],y=tab[0][r&127]^tab[1][(r>>7)&127]^tab[2][(r>>14)&127]^tab[3][r>>21];
      need(y>=r,"literal full-permutation canonical representative");
      stabilizers[i]+=(y==r);
    }
  } while(std::next_permutation(perm.begin(),perm.end()));
  for(int i=0;i<q;++i) need(int64_t(sizes[i])*stabilizers[i]==permutations,"literal full-permutation orbit size");
  std::vector<U> triangles;
  for(int i=0;i<n;++i) for(int j=i+1;j<n;++j) for(int k=j+1;k<n;++k) triangles.push_back(compress((U(1)<<idx[i][j])^(U(1)<<idx[i][k])^(U(1)<<idx[j][k])));
  M weighted(q,V(q)),psi(q,V(q));
  for(int i=0;i<q;++i) for(U t:triangles) ++weighted[i][primal.label[compressed[i]^t]];
  for(U x=0;x<count;++x) {
    V row(q); for(U t:triangles) ++row[primal.label[x^t]];
    need(row==weighted[primal.label[x]],"all labelled rows match representatives");
  }
  for(int i=0;i<q;++i) for(int j=0;j<q;++j) need(int64_t(sizes[i])*weighted[i][j]==int64_t(sizes[j])*weighted[j][i],"multiplicity reversibility");
  V complement(q),dual_sizes(q),dual_reps(q),eigenvalues(q);
  for(int j=0;j<q;++j) {
    U seed=dual.members[j][0]; dual_reps[j]=seed; dual_sizes[j]=dual.members[j].size();
    complement[j]=dual.label[seed^(count-1)];
    for(U t:triangles) eigenvalues[j]+=1-2*(__builtin_popcount(seed&t)&1);
    for(int i=0;i<q;++i) for(U h:dual.members[j]) psi[i][j]+=1-2*(__builtin_popcount(h&compressed[i])&1);
  }
  std::cout<<"{\"n\":"<<n<<",\"states\":"<<count<<",\"literal_permutations\":"<<permutations<<",\"reps\":"; array_json(reps);
  std::cout<<",\"sizes\":"; array_json(sizes);
  std::cout<<",\"dual_reps\":"; array_json(dual_reps);
  std::cout<<",\"dual_sizes\":"; array_json(dual_sizes);
  std::cout<<",\"complement\":"; array_json(complement);
  std::cout<<",\"eigenvalues\":"; array_json(eigenvalues);
  std::cout<<",\"M\":"; matrix_json(weighted);
  std::cout<<",\"psi\":"; matrix_json(psi); std::cout<<"}\n";
}

File: 04-computation/eulerian_n8_fourier_stress_overnight_hexagon_sep05.py
#!/usr/bin/env python3
"""Exact compiled n8 stress, with independent Python full-universe controls.

Temporary compilation is local and disposable. All semantic checks survive -O.
A nonzero modular determinant certifies a rational determinant is nonzero;
no inference is made from a zero modular determinant alone.
"""
import hashlib
import json
from pathlib import Path
import subprocess
import tempfile

import eulerian_boolean_fourier_repair_overnight_hexagon_sep05 as previous


def require(ok, message):
    if not ok:
        raise RuntimeError(message)


def digest(value):
    return hashlib.sha256(json.dumps(value, sort_keys=True, separators=(",", ":")).encode()).hexdigest()


def modular_det(a, prime):
    require(all(len(row) == len(a) for row in a), "square modular matrix")
    b = [[x % prime for x in row] for row in a]
    det = 1
    for j in range(len(b)):
        pivot = next((i for i in range(j, len(b)) if b[i][j]), None)
        if pivot is None:
            return 0
        if pivot != j:
            b[pivot], b[j] = b[j], b[pivot]
            det = -det
        value = b[j][j]
        det = det * value % prime
        inverse = pow(value, -1, prime)
        for i in range(j + 1, len(b)):
            multiplier = b[i][j] * inverse % prime
            if multiplier:
                for k in range(j + 1, len(b)):
                    b[i][k] = (b[i][k] - multiplier * b[j][k]) % prime
            b[i][j] = 0
    return det % prime


def matrix_product(a, b):
    cols = list(zip(*b))
    return [[sum(x*y for x, y in zip(row, col)) for col in cols] for row in a]


def independent_control(data):
    n = data["n"]
    reps, orbit, sizes, index = previous.native.generator_orbits(n)
    orbits, complement, gauge = previous.switching_orbits(n)
    compressed = [sum(((r >> index[e]) & 1) << j for j, e in enumerate(gauge)) for r in reps]
    psi = [[sum(1-2*((h & x).bit_count() % 2) for h in members) for members in orbits] for x in compressed]
    triangles = previous.native.old.simple_cycles(n, 3, index)
    weighted = previous.native.old.weighted_operator(reps, orbit, triangles)
    require(data["reps"] == reps and data["sizes"] == sizes, "independent primal representatives and sizes")
    require(data["dual_reps"] == [min(o) for o in orbits] and data["dual_sizes"] == list(map(len, orbits)), "independent dual representatives and sizes")
    require(data["M"] == weighted and data["psi"] == psi and data["complement"] == complement, "independent full matrices and complement")


def main():
    source = Path(__file__).with_suffix(".cpp")
    reports = []
    primes = (101, 1009, 1000003)
    # Literal determinant positive/hostile controls, including nonzero integers
    # whose reduction at one prime vanishes.
    for p in primes:
        require(modular_det([[2, 3], [5, 7]], p) == p-1, "negative determinant control")
        require(modular_det([[1, 2], [2, 4]], p) == 0, "singular control")
    require(modular_det([[101]], 101) == 0 and modular_det([[101]], 1009) == 101, "modular-zero is not rational-zero")
    with tempfile.TemporaryDirectory(prefix="math-n8-fourier-") as directory:
        binary = str(Path(directory) / "enumerate")
        subprocess.run(["g++", "-O3", "-std=c++17", "-Wall", "-Wextra", str(source), "-o", binary], check=True)
        for n in range(3, 9):
            data = json.loads(subprocess.check_output([binary, str(n)], text=True))
            if n <= 7:
                independent_control(data)
            reps, psi, complement, weighted = (data[k] for k in ("reps", "psi", "complement", "M"))
            q = len(reps)
            require(q == previous.native.fixed.OrbitSampler(n).orbit_count, "independent Burnside count")
            require(sum(data["sizes"]) == sum(data["dual_sizes"]) == data["states"] == 2**((n-1)*(n-2)//2), "complete labelled universes")
            require(all(complement[complement[j]] == j for j in range(q)), "complement involution")
            even = [i for i, r in enumerate(reps) if r.bit_count() % 2 == 0]
            odd = [i for i, r in enumerate(reps) if r.bit_count() % 2]
            fixed = [j for j, k in enumerate(complement) if j == k]
            pairs = [(j, k) for j, k in enumerate(complement) if j < k]
            require(len(fixed) == len(even)-len(odd) and len(pairs) == len(odd), "parity/Fourier dimensions")
            require(all(psi[i][j] == 0 for i in odd for j in fixed), "selfdual modes even-supported")
            require(all(psi[i][a] == (-1)**reps[i].bit_count()*psi[i][b] for i in range(q) for a,b in pairs), "free-pair parity relation")
            # Independent character eigenvalue check uses the literal full matrix.
            product = matrix_product(weighted, psi)
            require(all(product[i][j] == data["eigenvalues"][j]*psi[i][j] for i in range(q) for j in range(q)), "complete Fourier diagonalization of multiplicity only")
            require(all(weighted[i][j] == 0 for side in (even, odd) for i in side for j in side), "native bipartition")
            c = [[int(weighted[i][j] > 0) for j in even] for i in odd]
            s = [[psi[i][j] for j in fixed] for i in even]
            pmat = [[psi[i][a]+psi[i][b] for a,b in pairs] for i in even]
            odd_basis = [[psi[i][a]-psi[i][b] for a,b in pairs] for i in odd]
            a = matrix_product(c, pmat)
            dets = {str(prime): modular_det(a, prime) for prime in primes}
            basis_dets = {str(prime): modular_det([ss+pp for ss,pp in zip(s,pmat)], prime) for prime in primes}
            odd_basis_dets = {str(prime): modular_det(odd_basis, prime) for prime in primes}
            require(any(basis_dets.values()), "complete even Fourier basis certified modulo a prime")
            require(any(odd_basis_dets.values()), "complete odd Fourier basis certified modulo a prime")
            # The test deliberately reports failure without declaring it a proof
            # of rational singularity; a positive value is an exact certificate.
            success = any(dets.values())
            report = {"n":n,"labelled_states":data["states"],"q":q,"even":len(even),"odd":len(odd),
                      "selfdual":len(fixed),"weighted_zero_modes":data["eigenvalues"].count(0),
                      "free_block_modular_determinants":dets,"even_basis_modular_determinants":basis_dets,
                      "odd_basis_modular_determinants":odd_basis_dets,
                      "free_block_nonsingular_certified":success,"full_data_sha256":digest(data),
                      "free_block_sha256":digest(a),"independent_full_matrix_python_control":n<=7,
                      "literal_permutation_representative_size_control":data["literal_permutations"],
                      "unrepaired_boolean_defect_rank":previous.exact_rank(previous.sp.Matrix(matrix_product(c,s))) if fixed else 0}
            if success:
                report.update(boolean_rank=2*len(odd), boolean_nullity=len(fixed))
            print("COMPLETE FOURIER STRESS", json.dumps(report,sort_keys=True),flush=True)
            reports.append(report)
    print("SEMANTIC SHA256",digest(reports))
    print("FINITE-EXACT n3..8; all-order transversality and native spectral gap remain OPEN")


if __name__ == "__main__":
    main()
